src/cli/cli_coap.cpp

```diff
@@ -0,0 +1,356 @@
+    }
+
+exit:
+
+    if (error != kThreadError_None && responseMessage != NULL)
+    {
+        sServer->OutputFormat("Cannot send CoAP response message: Error %d\r\n", error);
+        otMessageFree(responseMessage);
+    }
+    else
+    {
+        if ((otCoapHeaderGetType(aHeader) == kCoapTypeConfirmable) || otCoapHeaderGetCode(aHeader) == kCoapRequestGet)
+        {
+            sServer->OutputFormat("CoAP response sent successfully!\r\n");
+        }
+    }
+
+    (void)aHeader;
+    (void)aMessage;
+    (void)aMessageInfo;
+}
+
+ThreadError Coap::ProcessClient(int argc, char *argv[])
+{
+    ThreadError error = kThreadError_None;
+    otMessage *message;
+    otMessageInfo messageInfo;
+    otCoapHeader header;
+
+    // Default parameters
+    char coapUri[] = "test";
+    otCoapType coapType = kCoapTypeNonConfirmable;
+    otCoapCode coapCode = kCoapRequestGet;
+    otIp6Address coapDestinationIp;
+
+    VerifyOrExit(argc > 0, error = kThreadError_InvalidArgs);
+
+    // CoAP-Code
+    if (strcmp(argv[0], "GET") == 0)
+    {
+        coapCode = kCoapRequestGet;
+    }
+    else if (strcmp(argv[0], "POST") == 0)
+    {
+        coapCode = kCoapRequestPost;
+    }
+    else if (strcmp(argv[0], "PUT") == 0)
+    {
+        coapCode = kCoapRequestPut;
+    }
+    else if (strcmp(argv[0], "DELETE") == 0)
+    {
+        coapCode = kCoapRequestDelete;
+    }
+    else
+    {
+        ExitNow(error = kThreadError_Parse);
+    }
+
+    // Destination IPv6 address
+    if (argc > 1)
+    {
+        SuccessOrExit(error = otIp6AddressFromString(argv[1], &coapDestinationIp));
+    }
+    else
+    {
+        ExitNow(error = kThreadError_InvalidArgs);
+    }
+
+    // CoAP-URI
+    if (argc > 2)
+    {
+        strcpy(coapUri, argv[2]);
+    }
+    else
+    {
+        ExitNow(error = kThreadError_InvalidArgs);
+    }
+
+    // CoAP-Type
+    if ((argc > 3) && (strcmp(argv[3], "Confirmable") == 0))
+    {
+        coapType = kCoapTypeConfirmable;
+    }
+
+    otCoapHeaderInit(&header, coapType, coapCode);
+    otCoapHeaderGenerateToken(&header, 2);
+    SuccessOrExit(error = otCoapHeaderAppendUriPathOptions(&header, coapUri));
+    otCoapHeaderSetPayloadMarker(&header);
+
+    message = otCoapNewMessage(sInstance, &header);
+    VerifyOrExit(message != NULL, error = kThreadError_Failed);
+
+    // Embed content into message if given
+    if (argc > 4)
+    {
+        SuccessOrExit(error = otMessageAppend(message, &argv[4], sizeof(argv[4])));
+    }
+
+    memset(&messageInfo, 0, sizeof(messageInfo));
+    messageInfo.mPeerAddr = coapDestinationIp;
+    messageInfo.mPeerPort = OT_DEFAULT_COAP_PORT;
+    messageInfo.mInterfaceId = OT_NETIF_INTERFACE_ID_THREAD;
+
+    error = otCoapSendRequest(sInstance, message, &messageInfo,
+                              (otCoapResponseHandler) &Coap::s_HandleClientResponse, NULL);
+
+    sServer->OutputFormat("Sending CoAP message: ");
+
+    if (error != kThreadError_None)
```

Comment:
You can initialize `message` to `NULL` upon declaration, then check if `message != NULL` before calling `otMessageFree()`.